File: DFS_BFS/542-nearestExit-DFS.py

from math import inf

def nearestExit(maze,entry):
    ans=inf
    rows=len(maze)
    cols=len(maze[0])
    def isExit(r,c):
        if r==0 or c==0 or r==rows-1 or c==cols-1:
            return maze[r][c]=='.'
        return False
    def dfs(r,c,d,seen):
        if isExit(r,c) and (r,c)!=tuple(entry):
            nonlocal ans
            ans=min(ans,d)
            return True
        if (r,c) in seen:
            return False
        
        seen.add((r,c))
        
        for dr,dc in ((0,1),(1,0),(-1,0),(0,-1)):
            nr,nc=dr+r,dc+c
            if nr>=0 and nr<=rows-1 and nc>=0 and nc<=cols-1 and maze[nr][nc]=='.':
                dfs(nr,nc,d+1,seen)
        seen.remove((r,c))
    seen=set()
    dfs(entry[0],entry[1],0,seen)
     
    return ans if ans<inf else -1

maze = [["+","+",".","+"],[".",".",".","+"],["+","+","+","."]]
entrance = [1,2]

print(nearestExit(maze, entrance))

File: DFS_BFS/542-nearestExit-DFS.cpp
#include<iostream>
#include<set>
#include<vector>

using std::cout;
using VEC=std::vector<char>;
using VEC_INT=std::vector<int>;
using MATRIX=std::vector<VEC>;

bool isExit(int r,int c,const MATRIX& m)
{
    int rows=m.size();
    int cols=m[0].size();
    if(r==0||c==0 ||r==rows-1||c==cols-1)
        return m[r][c]=='.';
    return false;
}

void dfs(const MATRIX& m,int r, int c,int d,std::set<VEC_INT>& seen,int &ans,const VEC_INT& entry)
{
    int rows=m.size();
    int cols=m[0].size();
    if(VEC_INT(r,c)!=entry &&(isExit(r,c,m)))
    {
        ans=std::min(ans,d);
        return ;
    }
    if(seen.find(VEC_INT(r,c))!=seen.end())
        return ;
    seen.insert(VEC_INT(r,c));
    std::vector<VEC_INT> dirs{{1,0},{0,1},{-1,0},{0,-1}};
    for(auto &dir:dirs)
    {
        int nr=dir[0]+r;
        int nc=dir[1]+c;
        if(nr>=0 &&nr<rows &&nc>=0 &&nc<cols &&m[nr][nc]=='.')
        {
            dfs(m,nr,nc,d+1,seen,ans,entry);
        }
    }
    seen.erase(VEC_INT(r,c));
}


int nearestExit(const MATRIX& m,  const VEC_INT& entry)
{
    int ans=INT_MAX;
    std::set<VEC_INT> seen{};
    dfs(m,entry[0],entry[1],0,seen,ans,entry);
    return ans!=INT_MAX?ans:-1;
}

int main()
{
    // maze = [['+','+','.','+'],['.','.','.','+'],['+','+','+','.']]
// entrance = [1,2]

    MATRIX m{{'+','+','.','+'},{'.','.','.','+'},{'+','+','+','.'}};
    VEC_INT entrance{1,2};
    cout<<nearestExit(m,entrance);
}